### PR TITLE
Refine logic for when curation removes a symbol from automatic curation

### PR DESCRIPTION
**Pull request by d-ronnqvist**
Bug/issue #, if applicable: rdar://133338975

## Summary

This refines the logic for when authored curation would remove a symbol from automatic curation. 

Before this, _any_ authored curation would remove a symbol or article from automatic curation.

With these changes, a symbol isn't removed from automatic curation if it's curated outside of its canonical container.

[This new code comment](https://github.com/swiftlang/swift-docc/pull/1075/files#diff-c452db10128fc43b2e77212a1a282fe0834e541e10852fa8dd50f7c9b480b4e7R2617-R2645) explains the new logic well:

> For symbols we only stop automatic curation if they are curated within their canonical container's sub-hierarchy or if a top-level symbol is curated under another top-level symbol (more on that below).
>
> For example, curating a member under an API collection within the container removes the member from automatic curation:
> ```
>  ┆
>  ├─SomeClass
>  │ └─API Collection
>  │   └─SomeClass/someMethod()  ◀︎━━ won't auto curate
> ```
> However, curating a member outside under another container _doesn't_ remove it from automatic curation:
> ```
>  ┆
>  ├─Other Container
>  │ └─SomeClass/someMethod()  ◀︎━━  will still auto curate under `SomeClass`
>  ├─SomeClass
> ```
> The same applies if the authored curation location is a another member of the canonical container:
> ```
>  ┆
>  ├─SomeClass
>  │ └─SomeClass/SomeInnerClass
>  │   └─SomeClass/someMethod()  ◀︎━━ will still auto curate under `SomeClass`
> ```
> Top-level symbols curated under other top-level is an exception to this rule.
> ```
>  ┆
>  ├─SomeClass
>  │ └─OtherTopLevelClass  ◀︎━━ won't auto curate because it's top-level.
> ```
> The reason for this exception is to allow developers to group top-level types under one-another without requiring an API collection. For example, in DocC one could curate `DiagnosticConsumer`, `DiagnosticFormattingOptions`, and `Diagnostic` under `DiagnosticEngine`, treating the `DiagnosticEngine` as the top-level topic for all diagnostic related types.

[These added tests](https://github.com/swiftlang/swift-docc/pull/1075/files#diff-691652dd8c4ae1ed9f9a0830248ce595b61cb289207d8e9d84c630b84faa947cR835-R1195) also verify when specific types curation do or don't remove a symbol from automatic curation.

## Dependencies

None.

## Testing

In a project with some symbols:
- Curate a top-level symbol under its module. 
  - This will stop automatically curating that symbol.
- Curate a top-level symbol under an API collection under the symbol's module. 
  - This will stop automatically curating that symbol.
- Curate a top-level symbol under another top-level symbol. 
  - This will stop automatically curating that symbol.
- Curate a top-level symbol under a _member_ of another symbol.
  - This won't stop automatically curating the top-level symbol.
 - Curate a member of some top-level symbol under _a different_ top-level symbol.
  - This won't stop automatically curating the member symbol. 
- Curate an article under a member of a container symbols.
  - This will stop automatically curating that article.

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- [x] Added tests
- [x] Ran the `./bin/test` script and it succeeded
- [x] Updated documentation if necessary

### CODE REVIEW

**Comment by d-ronnqvist** [comment]
@swift-ci please test

File: Sources/docc/DocCDocumentation.docc/adding-structure-to-your-documentation-pages.md [hunk 1]
```diff
@@ -206,6 +210,11 @@ documentation hierarchy.
 
 After you arrange nested symbols in an extension file, use DocC to compile your changes and review them in your browser.
 
+> Note: 
+> If you organize a member symbol into a topic group outside of its canonical container symbol, 
+> DocC will still include the member symbol in the default topic group.
+> This ensures that the reader can always find the member symbol somewhere in the sub-hierarchy of its canonical container symbol.
```
Comment (anferbui) [inline Sources/docc/DocCDocumentation.docc/adding-structure-to-your-documentation-pages.md:216]: An example here might be helpful, as I'm not sure how recognisable the term "canonical container symbol" will be for those reading this documentation.

**Review by anferbui — APPROVED** [review]
Looks good to me, left some small comments.

File: Sources/SwiftDocC/Infrastructure/DocumentationContext.swift [hunk 2]
```diff
@@ -2409,7 +2409,7 @@ public class DocumentationContext {
     ///
     /// - Parameter automaticallyCurated: A list of automatic curation records.
     func removeUnneededAutomaticCuration(_ automaticallyCurated: [AutoCuratedSymbolRecord]) {
-        // It might look like it would be correct to check `topicGraph.nodes[symbol]?.isManuallyCurated` here,
+        // It might look like it would be correct to check `topicGraph.nodes[symbol]?.isManuallyCuratedInContainer` here,
```
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/DocumentationContext.swift:2412]: Yes, thank you. I renamed that property more than once for this.

[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/DocumentationContext.swift]: Fixed in https://github.com/swiftlang/swift-docc/pull/1075/commits/f4265d3d79ef64898a309992ec046e5cb3d17d16

File: Sources/docc/DocCDocumentation.docc/adding-structure-to-your-documentation-pages.md (hunk 1 above)
Comment (d-ronnqvist) [reply Sources/docc/DocCDocumentation.docc/adding-structure-to-your-documentation-pages.md]: I rephrased this in https://github.com/swiftlang/swift-docc/pull/1075/commits/f78895066bdbde0b74cad6c5bf1087e60eb302bf to "the type that defines the member". That sentence gets a bit confusing with symbols that have different locations in different language representations but I think overall fewer people will be confused by it.

**Comment by d-ronnqvist** [comment]
@swift-ci please test

**Pull request merged**